src/web.rs

```diff
@@ -800,6 +883,50 @@ impl Service {
                                                                  e))))
     }
 
+    /// Returns a future separating the request from its form body.
+    ///
+    /// If this is not a `POST` or the body's `Content-Type` is not
+    /// `application/json`, returns an appropriate error response instead.
+    ///
+    /// Use with `and_then` to chain logic which consumes the json body.
+    fn with_json_body(&self, mut req: Request<hyper::Body>)
+                      -> Box<Future<Item = (Request<hyper::Body>, serde_json::Value),
```

Comment:
I think you could have the second part be a `String`, so you don't have to convert through the `json::Value` intermediary. The chained future would call `serde_json::from_str()` instead of `serde_json::from_value()` then.